Sources/SPMBuildCore/PluginInvocation.swift

```diff
@@ -18,6 +18,144 @@ import TSCUtility
+    /// determine the context that is available to the plugin.
+    ///
+    /// The working directory should be a path in the file system into which the plugin is allowed to write information
+    /// that persists between all invocations of a plugin for the same purpose. The exact meaning of "same" means here
+    /// depends on the particular plugin; for a build tool plugin, it might be the combination of the plugin and target
+    /// for which it is being invoked.
+    ///
+    /// Note that errors thrown by this function relate to problems actually invoking the plugin. Any diagnostics that
+    /// are emitted by the plugin are contained in the returned result structure.
+    ///
+    /// - Parameters:
+    ///   - action: The plugin action (i.e. entry point) to invoke, possibly containing parameters.
+    ///   - package: The root of the package graph to pass down to the plugin.
+    ///   - scriptRunner: Entity responsible for actually running the code of the plugin.
+    ///   - outputDirectory: A directory under which the plugin can write anything it wants to.
+    ///   - toolNamesToPaths: A mapping from name of tools available to the plugin to the corresponding absolute paths.
+    ///   - fileSystem: The file system to which all of the paths refers.
+    ///
+    /// - Returns: A PluginInvocationResult that contains the results of invoking the plugin.
+    public func invoke(
+        action: PluginAction,
+        package: ResolvedPackage,
+        buildEnvironment: BuildEnvironment,
+        scriptRunner: PluginScriptRunner,
+        outputDirectory: AbsolutePath,
+        toolNamesToPaths: [String: AbsolutePath],
+        observabilityScope: ObservabilityScope,
+        fileSystem: FileSystem
+    ) throws -> PluginInvocationResult {
+        // Create the plugin working directory if needed (but don't do anything with it if it already exists).
+        do {
+            try fileSystem.createDirectory(outputDirectory, recursive: true)
+        }
+        catch {
+            throw PluginEvaluationError.outputDirectoryCouldNotBeCreated(path: outputDirectory, underlyingError: error)
+        }
+
+        // Create the input context to send to the plugin.
+        // TODO: Some of this could probably be cached.
+        var serializer = PluginScriptRunnerInputSerializer(buildEnvironment: buildEnvironment)
+        let inputStruct = try serializer.makePluginScriptRunnerInput(
+            rootPackage: package,
+            pluginWorkDir: outputDirectory,
+            builtProductsDir: outputDirectory,  // FIXME — what is this parameter needed for?
+            toolNamesToPaths: toolNamesToPaths,
+            pluginAction: action)
+        
+        // Serialize the PluginEvaluationInput to JSON.
+        let encoder = JSONEncoder()
+        encoder.outputFormatting = [.prettyPrinted, .sortedKeys, .withoutEscapingSlashes]
```

Comment:
can we use `JSONEncoder.makeWithDefaults()` ?

Reply:
Yes, definitely.  Great point.